Don't overwrite pinned assembly versions in servicing

ViktorHofer opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/84031

Eight packable projects pin their assembly version for .NET Framework compatibility. The incremental package servicing infrastructure didn't check if the assembly version is pinned and changed it during servicing.

As an example, System.Speech pins its assembly version to 4.0.0.0 but that version gets overwritten during servicing. I.e. for .NET 7 the version would then change to "7.0.0.$(ServicingVersion)" which is incorrect.

Please find the full list of impacted assemblies below:
- System.ComponentModel.Composition
- System.DirectoryServices
- System.DirectoryServices.AccountManagement
- System.DirectoryServices.Protocols
- System.Management
- System.Reflection.Context
- System.Runtime.Caching
- System.Speech

This change should be backported into release/6.0 and release/7.0.

ViktorHofer committed:
Don't overwrite pinned assembly versions in servicing

There are eight packable projects that pin their assembly version for .NET
Framework compatibility. The incremental package servicing
infrastructure didn't check if the assembly version is pinned and always
changed it.

As an example, System.Speech pins its assembly version to 4.0.0.0 but
that version gets overwritten during servicing. I.e. for .NET 7 the
version would then change to "7.0.0.$(ServicingVersion)" which is
incorrect.

Please find the full list of impacted assemblies below:
- System.ComponentModel.Composition
- System.DirectoryServices
- System.DirectoryServices.AccountManagement
- System.DirectoryServices.Protocols
- System.Management
- System.Reflection.Context
- System.Runtime.Caching
- System.Speech

ViktorHofer commented on src/libraries/System.Reflection.Context/Directory.Build.props:
```diff
@@ -1,6 +1,9 @@
+    <!-- this assembly is inbox in desktop, do not version it unless you
+         plan on shipping a new desktop version out of band. Instead add API
+         to a different assembly. -->
     <AssemblyVersion>4.0.3.0</AssemblyVersion>
```
@ericstj do you know why this library is versioned higher than its .NET Framework runtime counterpart?

 > "C:\Windows\Microsoft.NET\Framework\v4.0.30319\System.Reflection.context.dll"

`[assembly: AssemblyVersion("4.0.0.0")]`

ericstj replied:
Maybe because it was OOBed?  

.NET's unification table will shave off the last two parts of the version so this reference will still unify to the 4.0.0.0 version at runtime.  Not sure what it does at compile time.  It might be handled by RAR's AutoUnify flag.

Oddly it was even 4.0.4.0 in 3.1 https://github.com/dotnet/corefx/blob/b58ef0a19eca6269799664d6288be14b7f0cef3e/src/System.Reflection.Context/Directory.Build.props#L4

ericstj replied:
As I mentioned in the other comment this same property was overridden in the ref-project to be set as `4.0.0.0`.  It looks like @ViktorHofer did this in https://github.com/dotnet/corefx/pull/29347.

At that time it was OK to do that, since we still shipped reference assemblies.  In 6.0 when we stopped shipping reference assemblies this higher version was exposed.

For the most part it's fine and doesn't hurt anything at runtime.  .NETFramework will still load this type just fine due to unification as I mentioned.

There is a warning from RAR on .NETFramework due to this mismatch, but it's just a warning.

We'd actually make things a bit worse if we went backwards on this version (at least in any runtime assembly). 

One thing we could consider -- if we merged this change in 8.0 to only pin the version for the `netstandard2.0` assembly - we could rollback the version to `4.0.0.0`.  That would fix the warning in the netframework/netstandard2.0 case, at the cost of introducing the same warning for people in the case of compiling against the old package in one library and using it in project referencing the new one.  Hard to say if it's worth the effort.  Regardless I'll open an issue for that: https://github.com/dotnet/runtime/issues/84320

ViktorHofer commented on src/libraries/System.Reflection.Context/ref/System.Reflection.Context.csproj:
```diff
@@ -1,9 +1,5 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <!-- this assembly is inbox in desktop, do not version it unless you
-         plan on shipping a new desktop version out of band. Instead add API
-         to a different assembly. -->
-    <AssemblyVersion>4.0.0.0</AssemblyVersion>
```
Removed as the library's Directory.Build.props already pins the AssemblyVersion.

ericstj replied:
As you mentioned above this one was pinned to `4.0.3.0` in the Directory.Build.props. So this will mean that project references will now see the higher versioned reference.  That could cause a problem if they were loaded on .NETFramework and for whatever reason unification wasn't working (design time). 

The only project that I see that did this was `System.ComponentModel.Composition.Registration.csproj`.  That doesn't ship a copy on .NETFramework, so it's likely not a problem.

ericstj replied:
Resolving this to have the conversation in the other thread https://github.com/dotnet/runtime/pull/84079#discussion_r1152273413

ViktorHofer commented on src/libraries/System.Data.DataSetExtensions/Directory.Build.props:
```diff
@@ -1,8 +1,6 @@
 ﻿<Project>
   <Import Project="..\Directory.Build.props" />
   <PropertyGroup>
-    <!-- AssemblyVersion is frozen to that which originally shipped in 2.1 -->
-    <AssemblyVersion>4.0.0.0</AssemblyVersion>
```
This library only ships inside the Microsoft.NETCore.App shared framework. I assume we don't need to pin its assembly version anymore?

ericstj replied:
I think it once had a package and this is left over from that time.

carlossanlop replied:
What is the expected side-effect of removing AssemblyVersion from here?

ericstj replied:
The assembly will version in the same way as every other assembly that ships only in the targeting packs (N.0.0.0 which remains frozen in servicing).

ericstj reviewed (commented):
Did you consider only applying AssemblyVersion to the `netstandard2.0` assembly -- that way we wouldn't break folks targeting .NETCore who consumed the serviced packages?  It also limits the "ugliness" of this pinned assembly version - letting us do the right thing in new versions and only having this fixed version for .NETStandard.

ericstj committed:
Only pin AssemblyVersion in netstandard2.0 assemblies

These assemblies pin their AssemblyVersion to be compatible with
.NETFramework where the same assembly is inbox.

The only TargetFramework we build that is compatible with .NETFramework
is netstandard2.0, so only pin the assembly on that framework.

ericstj commented:
I tested this change in servicing and noticed that it isn't behaving quite right.
https://gist.github.com/ericstj/308e0ba41e977fe7c4f6e1025d076e0e

ericstj commented on eng/packaging.targets:
```diff
@@ -48,8 +48,10 @@
+    <!-- The assembly version gets updated when
+      - it isn't already set (i.e. it could be pinned to a specific version for .NET Framework compatibility) and
+      - the assembly isn't part of a targeting pack (always true for .NET Framework). -->
+    <AssemblyVersion Condition="'$(AssemblyVersion)' == '' and '$(_AssemblyInTargetingPack)' != 'true'">$(MajorVersion).$(MinorVersion).0.$(ServicingVersion)</AssemblyVersion>
```
This isn't setting `AssemblyVersion` if it's already set, but it looks like it will *always* be set, because it's done here: https://github.com/dotnet/runtime/blob/8b1d1eabe32ba781ffcce2867333dfdc53bdd635/eng/Versions.props#L18

A few other approaches:
1. A flag representing that a library controls its own AssemblyVersion
2. A property like `FixedAssemblyVersion`
3. Remove this condition entirely and just move the place where `AssemblyVersion` is hardcoded in the individual libraries to `Directory.Build.targets` after this is evaluated.

I'm inclined to do the latter.

carlossanlop replied:
A few questions:

> A flag representing that a library controls its own AssemblyVersion

With this approach, we would substitute the `'$(AssemblyVersion)' == ''` condition with `'$(AssemblyControlsOwnVersion)' == 'true'`, corect? And the property would be enabled in each csproj that needs it, I assume. Would that mean we woulc have to resort to using hardcoded numbers instead of using `$(MajorVersion).$(MinorVersion).0.$(ServicingVersion)`?

> A property like FixedAssemblyVersion

Does this mean we would have to update all our OOB packages to use this new property instead of AssemblyVersion?

> and just move the place where AssemblyVersion is hardcoded in the individual libraries to Directory.Build.targets after this is evaluated.

Do you mind elaborating on this? I'm not sure I understand this option which you prefer. Do you mean having a centralized place (Directory.Build.targets) to control the OOB package versions?

ericstj replied:
`AssemblyControlsOwnVersion` - yeah something like that.  It would only be used in these 8 projects that need to set it.  Putting it in the project would tell `packaging.targets` not to overwrite it with it.

`FixedAssemblyVersion` -- only these 8 projects.  This would just be a single property rather than `AssemblyControlsOwnVersion` and `AssemblyVersion`.  We'd then have some shared targets file that would set `AssemblyVersion` to `FixedAssemblyVersion` and it would do so in/after packaging.targets.

The third option is just moving the setting of `AssemblyVersion` out of the library-specific `Directory.Build.props` into a library-specific `Directory.Build.targets`.  Here's that change, still testing out on servicing branches: https://github.com/ericstj/runtime/commit/3a09813b43361181f47c5b15f11bc1a9e79c8711

This is all about order of evaluation and what state needs to be communicated.

ericstj committed:
Override AssemblyVersion in Directory.Build.targets

Previously assemblies would set AssemblyVersion in Directory.Build.props
which was too early, because later packaging.targets could not distinguish
between a value that was explicitly set and the default value of the repo.

Instead override the value after the Servicing value is set, in
Directory.Build.targets

ericstj commented:
Of the 8 packages listed, `System.DirectoryServices` is also part of the WindowsDesktop shared framework, so that prevented it from hitting this problem.

While I do think we should make the change I mentioned above https://github.com/dotnet/runtime/pull/84079#pullrequestreview-1363707444 to reduce the places where we hardcode the version, I recommend we avoid doing that in servicing for `System.DirectoryServices` since it isn't broken and that change would leak into the targeting pack.

carlossanlop commented on src/libraries/System.Reflection.Context/Directory.Build.props:
```diff
@@ -1,7 +1,10 @@
+    <!-- this assembly is inbox in desktop, do not version it unless you
+         plan on shipping a new desktop version out of band. Instead add API
+         to a different assembly. -->
+    <AssemblyVersion Condition="'$(TargetFramework)' == 'netstandard2.0'">4.0.3.0</AssemblyVersion>
```
Why is this not "4.0.0.0"? Does "4.0.3.0" mean we serviced it in the past?

ericstj replied:
resolving this, since that conversation is happening above in https://github.com/dotnet/runtime/pull/84079#discussion_r1152273413.

ericstj committed:
Update comments

ericstj committed:
Remove ValidateAssemblyVersionsInRefPack target

The only thing this target did was a redundant check that AssemblyVersion was
set for projects in the Targeting Pack.  That check was failing for
the NETStandard build of System.DirectoryServices.

This target was providing no value, instead we need to fix https://github.com/dotnet/runtime/issues/42961
which will be a proper check of our servicing version rules.

ericstj commented:
Here's the 7.0 version of the fix. https://github.com/dotnet/runtime/compare/release/7.0-staging...ericstj:runtime:FixPinnedAssemblyVersion-7.0?expand=1

Here's the diff of assembly versions: https://gist.github.com/ericstj/5e914be725066df9824fd1e3fca12c08

Note the change for 5 binaries which we never shipped.  These were still seeing the `7.0.0.0` version because their version was being replaced with a `ServicingVersion` of 0.  We would have noticed this problem if we shipped any other packages which had references to these projects.  I don't think we've done this but I'll audit all references and update this comment.

carlossanlop commented on src/libraries/System.Runtime.Caching/Directory.Build.targets:
```diff
@@ -1,9 +1,8 @@
-         plan on shipping a new desktop version out of band. Instead add API
-         to a different assembly. -->
+    <!-- This assembly is inbox in .NETFramework, ensure that the .NETStandard assembly
+         remains <= the desktop version -->
```
Tiny nit: I see desktop was changed to .NETFramework in line 4, but not in line 5. Not asking to retrigger the CI just for this, but maybe the same language should be applied here.

ericstj replied:
I'd prefer to get it right.

carlossanlop commented on src/libraries/System.Management/Directory.Build.targets:
```diff
@@ -0,0 +1,8 @@
+  <PropertyGroup>
+    <!-- This assembly is inbox in .NETFramework, ensure that the .NETStandard assembly
+         remains <= the desktop version -->
+    <AssemblyVersion Condition="'$(TargetFramework)' == 'netstandard2.0'">4.0.0.0</AssemblyVersion>
```
Ok nice. I can see that you're shipping a new version of System.Management in the backports.

carlossanlop reviewed (approved):
LGTM. If the manual tests work, let's merge it.
And thanks for sending the backports.

ericstj committed:
Address feedback

ViktorHofer reviewed (commented):
Thanks, that looks great. Unfortunately I didn't notice that the AssemblyVersion property was already set in Versions.props.
